plugins/outputs/loki/loki.go

```diff
@@ -0,0 +1,214 @@
+	Scopes       []string          `toml:"scopes"`
+	GZipRequest  bool              `toml:"gzip_request"`
+
+	url    string
+	client *http.Client
+	tls.ClientConfig
+}
+
+func (l *Loki) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *Loki) Description() string {
+	return "Send logs to Loki"
+}
+
+func (l *Loki) createClient(ctx context.Context) (*http.Client, error) {
+	tlsCfg, err := l.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("tls config fail: %w", err)
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: l.Timeout.Duration,
+	}
+
+	if l.ClientID != "" && l.ClientSecret != "" && l.TokenURL != "" {
+		oauthConfig := clientcredentials.Config{
+			ClientID:     l.ClientID,
+			ClientSecret: l.ClientSecret,
+			TokenURL:     l.TokenURL,
+			Scopes:       l.Scopes,
+		}
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
+		client = oauthConfig.Client(ctx)
+	}
+
+	return client, nil
+}
+
+func (l *Loki) Connect() (err error) {
+	if l.Domain == "" {
+		return fmt.Errorf("domain is required")
+	}
+
+	if l.Endpoint == "" {
+		l.Endpoint = defaultEndpoint
+	}
+
+	l.url = fmt.Sprintf("%s%s", l.Domain, l.Endpoint)
+
+	if l.Timeout.Duration == 0 {
+		l.Timeout.Duration = defaultClientTimeout
+	}
+
+	ctx := context.Background()
+	l.client, err = l.createClient(ctx)
+	if err != nil {
+		return fmt.Errorf("http client fail: %w", err)
+	}
+
+	return
+}
+
+func (l *Loki) Close() error {
+	l.client.CloseIdleConnections()
+
+	return nil
+}
+
+func (l *Loki) Write(metrics []telegraf.Metric) error {
+	s := Streams{}
+
+	for _, m := range metrics {
+		tags := m.TagList()
+		var line string
+
+		for _, f := range m.FieldList() {
+			line += fmt.Sprintf("%s=\"%v\" ", f.Key, f.Value)
+		}
+
+		s.insertLog(tags, Log{fmt.Sprintf("%d", m.Time().UnixNano()), line})
+	}
+
+	return l.write(s)
+}
+
+func (l *Loki) write(s Streams) error {
+	bs, err := json.Marshal(s)
+	if err != nil {
+		return fmt.Errorf("json.Marshal: %w", err)
+	}
+
+	var reqBodyBuffer io.Reader = bytes.NewBuffer(bs)
+
+	if l.GZipRequest {
+		rc, err := internal.CompressWithGzip(reqBodyBuffer)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		reqBodyBuffer = rc
+	}
+
+	req, err := http.NewRequest(http.MethodPost, l.url, reqBodyBuffer)
+	if err != nil {
+		return err
+	}
+
+	if l.Username != "" {
+		req.SetBasicAuth(l.Username, l.Password)
+	}
+
+	for k, v := range l.Headers {
+		if strings.ToLower(k) == "host" {
+			req.Host = v
+		}
+		req.Header.Set(k, v)
+	}
+
+	req.Header.Set("User-Agent", internal.ProductToken())
+	req.Header.Set("Content-Type", "application/json")
+	if l.GZipRequest {
+		req.Header.Set("Content-Encoding", "gzip")
+	}
+
+	resp, err := l.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
```

Comment:
maybe check the body after the response status. There are plenty of response statuses that don't return bodies, and the Go library no longer leaks if you don't read the body.

Reply:
I think the ReadAll is useless here, the body isn't used. I've remove it and only check the status code